### PR TITLE
St/improved train

### PR DESCRIPTION
**Pull request by SamTallo01**

### CODE REVIEW

File: train/train.py [hunk 1]
```diff
@@ -17,131 +19,180 @@
...
+            else:
+                loss = harmoniccnn_loss(
+                    yo_pred,
+                    yp_pred,
+                    yo_true_batch,
+                    yp_true_batch,
+                    label_smoothing=s.label_smoothing,
+                    weighted=s.weighted,
+                    positive_weight=s.positive_weight,
                 )
```
Comment (FrancescoRisso) [inline train/train.py:121]: Idem come sopra

File: train/train.py [hunk 2]
```diff
@@ -17,131 +19,180 @@
...
+                    "yp_pred": yp_pred_sig.detach().cpu(),
+                    "yn_pred": yn_pred_sig.detach().cpu() if not s.remove_yn else None,
+                    "yo_true": yo_true_batch.detach().cpu(),
+                    "yp_true": yp_true_batch.detach().cpu(),
+                    "yn_true": (
+                        yn_true_batch.detach().cpu() if not s.remove_yn else None
+                    ),
+                }
+
+        else:
```
Comment (FrancescoRisso) [inline train/train.py:137]: Anche qui aggiungimi un TODO

File: train/train.py [hunk 3]
```diff
@@ -150,20 +201,103 @@ def train():
...
+                        example_outputs["yn_true"][0]
+                        if example_outputs["yn_true"] is not None
+                        else None
+                    ),
+                )
+                wandb.log(
+                    {"prediction_vs_gt": wandb.Image(fig, caption=f"Epoch {epoch+1}")},
+                    step=epoch + 1,
+                )
+                plt.close(fig)
```
Comment (FrancescoRisso) [inline train/train.py:286]: Anche qui aggiungimi un else TODO

File: train/utils.py [hunk 4]
```diff
@@ -60,167 +61,232 @@ def to_tensor(array):
...
+#             DataSet(Split.TRAIN, s.seconds), batch_size=s.batch_size, shuffle=True
+#         ),
+#     )
+
+
+# @torch.no_grad()
+# def run_single_batch_prediction_plot(
+#     model_path: str,
+#     model: torch.nn.Module,
+#     dataloader: torch.utils.data.DataLoader,
```
Comment (FrancescoRisso) [inline train/utils.py:230]: Togli tutta sta parte commentata (anche sotto)

File: .gitignore [hunk 5]
```diff
@@ -181,4 +181,10 @@ data_1/
...
 # model saves
-model_saves/
\ No newline at end of file
+model_saves/
+
+# wandb
+wandb/
+
+# audio for plotting 
+shouldbesame.wav
```
Comment (FrancescoRisso) [inline .gitignore:190]: Nah questo dovrebbe essere generato solo dal main di prova per provare il dataset singolo
Cancella pure il file e questa entry qui lol

File: settings.py [hunk 6]
```diff
@@ -42,20 +42,27 @@ class Settings:
...
     weighted = True
+    positive_weight_yp = 0.55
+    positive_weight_yo = 0.90
+    positive_weight_yn = 0.5
     save_model = True
 
+    # Training on single element
+    single_element_training = (
+        False  # if True, the model will be trained on a single element
+    )
```
Comment (FrancescoRisso) [inline settings.py:64]: Posso dire che sto coso è bruttissimo?
Piuttosto metti il commento alla riga sopra, ma cerca di fargli togliere ste parentesi su tre righe per un banalissimo False 😆

File: train/losses.py [hunk 7]
```diff
@@ -20,76 +20,64 @@ def weighted_transcription_loss(
...
+
+    results = {"loss_yo": loss_onset, "loss_yp": loss_tone}
+
+    if yn_logits is not None and yn_true is not None:
+        loss_note = weighted_transcription_loss(
+            yn_true, yn_logits, label_smoothing, positive_weight_yn
+        )
+        results["loss_yn"] = loss_note
+
+    return results
```
Comment (FrancescoRisso) [inline train/losses.py:83]: Manca newline

File: train/train.py [hunk 8]
```diff
@@ -150,20 +164,137 @@ def train():
     print(f"Training on {device}")
 
     timestamp = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
-    session_dir = os.path.join(f"model_saves", f"training_{timestamp}")
+    session_dir = os.path.join("model_saves")
+
+    wandb.init(
+        project="MLVM-Project",
+        name=f"harmonic_training_{timestamp}",
```
Comment (FrancescoRisso) [inline train/train.py:171]: Mi turba questo nome, per due motivi:
- è molto lungo
- non sai che modello è

Lo modificherei in qualcosa del tipo "{R|C}NN-{dataora}"

Cioè è vero che il modello è scritto nei config, ma se a colpo d'occhio voglio sapere come sono andati i modelli di un tipo rispetto all'altro, è complicato capire chi è chi.

File: train/utils.py [hunk 9]
```diff
@@ -60,167 +58,169 @@ def to_tensor(array):
...
+
+    # Plot YN (solo se non rimosso)
+    if not s.remove_yn:
+        im4 = axes[2, 0].imshow(yn_true_np, aspect="auto", origin="lower")
+        axes[2, 0].set_title("YN Ground Truth")
+        fig.colorbar(im4, ax=axes[2, 0])
+
+        im5 = axes[2, 1].imshow(yn_pred_np, aspect="auto", origin="lower")
+        axes[2, 1].set_title("YN Prediction")
+        fig.colorbar(im5, ax=axes[2, 1])
```
Comment (FrancescoRisso) [inline train/utils.py:149]: Domanda, si riuscirebbe a fare in modo che la scala di colori vada sempre da 0 a 1, in modo da avere grafici coerenti nel tempo?
Forse si può fare qualcosa [tipo questo](https://www.reddit.com/r/learnpython/comments/17089oo/matplotlib_force_colorbar_to_show_range_that_is/).

**Review by FrancescoRisso — CHANGES_REQUESTED** [review]

**Comment by FrancescoRisso** [comment]
Also, fix l'accuracy negativa 😆

File: train/evaluate.py [hunk 10]
```diff
@@ -45,10 +38,9 @@ def evaluate(model_path, dataset):
     test_loader = DataLoader(test_dataset, batch_size=s.batch_size, shuffle=False)
 
     running_loss = 0.0
+    running_acc_yp = 0.0
```
Comment (FrancescoRisso) [inline train/evaluate.py:41]: Non so se sia il posto giusto, ma fixa l'accuracy negativa lol

File: train/train.py [hunk 11]
```diff
@@ -150,20 +164,137 @@ def train():
...
+                        "metrics_VAL/average accuracy/train_yp": val_metrics[
+                            "yp_accuracy"
+                        ],
+                        "metrics_VAL/precision/train_yp": val_metrics["yp_precision"],
+                        "metrics_VAL/recall/train_yp": val_metrics["yp_recall"],
+                        "metrics_VAL/f1/train_yp": val_metrics["yp_f1"],
+                        "metrics_VAL/TP/train_yp": val_metrics["yp_TP"],
+                        "metrics_VAL/FP/train_yp": val_metrics["yp_FP"],
+                        "metrics_VAL/FN/train_yp": val_metrics["yp_FN"],
+                        "metrics_VAL/TN/train_yp": val_metrics["yp_TN"],
```
Comment (FrancescoRisso) [inline train/train.py:266]: TP/TN/FP/FN sono utili per calcolare precision e recall, ma penso che i loro valori puri non siano granchè indicativi (es TP è basso cosa vuol dire? Ne ha trovati pochi lui? O li ha trovati tutti, ma in questo giro erano pochi loro?)
Quindi li lascerei nel calcolo (perchè servono per gli altri lol), ma non li pusherei su wandb.

Poi questa è solo la mia opinione eh 😅

**Review by FrancescoRisso — CHANGES_REQUESTED** [review]

File: train/utils.py [hunk 12]
```diff
@@ -60,167 +58,169 @@ def to_tensor(array):
...
+            "Precision": precision,
+            "Recall": recall,
+            "F1": f1,
+            "Accuracy": accuracy,
+        }
+
+
+def plot_prediction_vs_ground_truth(
+    yo_pred, yp_pred, yn_pred, yo_true, yp_true, yn_true
+):
```
Comment (FrancescoRisso) [inline train/utils.py:108]: Se non è un problema, modificherei l'immagine per avere solo uno dei due (ground truth/output)
Così riusciamo a sparare su wandb il ground truth solo una volta, e l'output tante volte, poi li affianchiamo dalla dashboard.
In questo modo penso che dovremmo usare meno disco su wandb, che ne abbiamo abbastanza ma non infinito (e mi piacerebbe anche loggare audio e midi, che vanno a fare altri dati)

**Review by FrancescoRisso — CHANGES_REQUESTED** [review]

File: settings.py (hunk 6 above)
Comment (SamTallo01) [reply settings.py:64]: E che se formatto usciva sta roba, non avevo capito era colpa del commento lol

File: train/train.py (hunk 11 above)
Comment (SamTallo01) [reply train/train.py:266]: Io li avrei pushati lo stesso, non so a me piace vederli peró effettivamente abbiamo precision e recall se vuoi li rimuovo

Comment (FrancescoRisso) [reply train/train.py:266]: @FedericoPag tu cosa dici?

Comment (SamTallo01) [reply train/train.py:266]: Comunque ci sta rimuoverli alla fine come ficevi pirma abbiamo memoria limitata

Comment (FrancescoRisso) [reply train/train.py:266]: Beh si per quello sono due numerini, non occupano granchè

Comment (SamTallo01) [reply train/train.py:266]: Vabbé dai io gli ho rimossi, se ci servono poi li rimetto non ci va niente

File: .gitignore [hunk 13]
```diff
@@ -184,7 +184,4 @@ audio-fonts/*
 model_saves/
 
 # wandb
-wandb/
-
-# audio for plotting 
-shouldbesame.wav
+wandb/
```
Comment (FrancescoRisso) [inline .gitignore:187]: Manca il newline 🤣

File: train/utils.py [hunk 14]
```diff
@@ -103,54 +103,62 @@ def binary_classification_metrics(
...
+    imshow_fixed(axes[1], yp_pred_np, "YP Prediction", fig)
+
+    if not s.remove_yn:
+        imshow_fixed(axes[2], yn_pred_np, "YN Prediction", fig)
+
+    for ax in axes:
+        ax.axis("off")
 
     plt.tight_layout()
     return fig
```
Comment (FrancescoRisso) [inline train/utils.py:164]: Sono letteralmente la stessa funzione, cambiando solo la label 😆 

Sinceramente ne farei una sola, così evitiamo la duplicazione di codice. 
Più che altro così se ci fosse da toccare qualcosa basta modificare una volta sola, non rischiamo di avere cose incoerenti.

Per il nome, ci sono due alternative: aggiungere un parametro stringa, oppure semplicemente chiamare solo "YO", "YP", "YN" i plot, tanto poi prediction/ground truth è già scritto nel titolo di wandb.

**Review by FrancescoRisso — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 13 above)
Comment (FedericoPag) [reply .gitignore:187]: Che sei fiscale ahahah

Comment (SamTallo01) [reply .gitignore:187]: Che palle sto newline HAHAHAHA

Il bello é che lo vado a cercare sempre e tutte le volte me lo perdo

Comment (FrancescoRisso) [reply .gitignore:187]: Ma basta formattare i file
Che tra l'altro, se hai il format-on-save basta fare ctrl-s quando passi ad un altro

Ma sono fiscale sia perchè è good practice, sia perchè mi turba il divieto d'accesso che la review di github mette se manca il newline lol

**Review by FedericoPag — APPROVED** [review]
